### PR TITLE
[ALLUXIO-3044] Classload HDFS UFS connectors

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-3044

### CODE REVIEW

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java [hunk 1]
```diff
@@ -153,28 +163,50 @@ public static UnderFileSystemFactory find(String path) {
   }
 
   /**
+   * Finds all {@link UnderFileSystemFactory} from the lib directory and caches.
+   */
+  private static void loadLibs() {
```
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: nit: alphabetically sort

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java [hunk 2]
```diff
@@ -153,28 +163,50 @@ public static UnderFileSystemFactory find(String path) {
...
+    List<File> files = new ArrayList<>();
+    try (DirectoryStream<Path> stream = Files.newDirectoryStream(
+        Paths.get(Configuration.get(PropertyKey.HOME), "lib"), "alluxio-underfs-*.jar")) {
```
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: seems like the lib directory might be used by non-underfs libs as well. can we define the location as a constant (not in ufs module) or a config parameter if it makes sense. well maybe not a config as i guess the build script will write to a constant path.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java [hunk 3]
```diff
@@ -153,28 +163,50 @@ public static UnderFileSystemFactory find(String path) {
...
+  /**
    * Finds all {@link UnderFileSystemFactory} extensions from the extensions directory and caches.
    */
-  private static void scanExtensions() {
-    for (File extension : ExtensionUtils.listExtensions()) {
+  private static void loadExtensions() {
+    scan(Arrays.asList(ExtensionUtils.listExtensions()), LOADED_EXTENSION_JARS);
+  }
+
+  private static void scan(List<File> files, Set<String> jars) {
```
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: nit: comment

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java (hunk 2 above)
Comment (gpang) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: there should also be a log message stating which dir is being searched for jars.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java (hunk 1 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: done

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java (hunk 2 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: done and done

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java (hunk 3 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: added

**Comment by apc999** [comment]
@gpang  PTAL.

**Pull request merged**